Fix: update lesson-note--variations styles a11y

luuu-xu opened the pull request:
<!-- Thank you for taking the time to contribute to The Odin Project. In order to get this pull request (PR) merged in a reasonable amount of time, you must complete this entire template. -->

## Because
<!-- Summarize the purpose or reasons for this PR, e.g. what problem it solves or what benefit it provides. -->

Current `lesson-note--warning` and `lesson-note--tip` etc has colours and icons not meeting good a11y standards.


## This PR
<!-- A bullet point list of one or more items describing the specific changes. -->

- `lesson-note` div now has vertical margins with `my-4`.
- `lesson-note--tip`'s colour is now **blue**
- `lesson-note--warning` now has **orange** colour instead of red, and uses a warning icon
- `lesson-note--critical` variation is added, with colour of **red**, and uses a critical icon
- All of lesson-note sections now will render an icon **in front of the first descendent** element (assuming `h1-h6`) with the appropriate font-size. 
- Otherwise if the section **only has one descendent** (assuming only `p`), the icon will be rendered at the top left corner, with an enlarged `font-size: 1.8rem`

## Issue
<!--
If this PR closes an open issue in this repo, replace the XXXXX below with the issue number, e.g. Closes #2013.

If this PR closes an open issue in another TOP repo, replace the #XXXXX with the URL of the issue, e.g. Closes https://github.com/TheOdinProject/curriculum/issues/XXXXX

If this PR does not close, but is related to another issue or PR, you can link it as above without the 'Closes' keyword, e.g. 'Related to #2013'.
-->
Closes #4045 

## Additional Information
<!-- Any other information about this PR, such as a link to a Discord discussion. -->


## Pull Request Requirements
<!-- Replace the whitespace between the square brackets with an 'x', e.g. [x]. After you create the PR, they will become checkboxes that you can click on. -->
-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `keyword: brief description of change` format, using one of the following keywords:
  - `Feature` - adds new or amends existing user-facing behavior
  - `Chore` - changes that have no user-facing value, refactors, dependency bumps, etc
  - `Fix` - bug fixes
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] I have verified all tests and linters pass after making these changes.
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [ ] If applicable, this PR includes new or updated automated tests

luuu-xu commented:
@KevinMulhern Hi Kevin please review this PR as well. It's a fix of styles only.

KevinMulhern commented:
Hey @luuu-xu, @thatblindgeye is best suited to review this one.

thatblindgeye commented:
@luuu-xu I'll be taking a look at this tomorrow morning, looking forward to seeing how things look!

thatblindgeye commented on app/javascript/stylesheets/lesson-content.css:
```diff
@@ -24,33 +24,83 @@
 
   .lesson-note--warning {
-    @apply border-red-700 dark:border-red-500;
+    @apply border-orange-700 dark:border-orange-600;
```
I feel like these shades of orange are a little too dark when compared to the critical color... What do you think about `@apply border-orange-500 dark:border-orange-400;` instead?

thatblindgeye reviewed (changes requested):
This looks awesome! I just have a couple of comments. In addition to the one here, there's a file comment below.

One thing I'm noticing is that the top padding for the note box is a little larger than the bottom padding when there's no heading:

![Note box showing larger top padding than bottom padding](https://github.com/TheOdinProject/theodinproject/assets/70952936/02791c0c-9362-4dc8-a365-03eb5a0f6738)

Not a major blocker, but it'd be nice if we could update this so that the spacing along the top of the note box is the same as the bottom.

luuu-xu commented:
> One thing I'm noticing is that the top padding for the note box is a little larger than the bottom padding when there's no heading:

Great eyes!! I have added `-mt-2` to these large icons to compensate the padding and it looks better now.

And I have updated the orange colours for `warning`, they do look much better now. I was too worried about colour contrasts where in fact icon and border does not need to comply strictly to contrast ratings.

thatblindgeye reviewed (approved):
![image](https://github.com/TheOdinProject/theodinproject/assets/70952936/581c0cdb-d7e8-4375-b726-5fd079af058d)

😍 

@KevinMulhern if there's nothing you want to look over from a Tailwind perspective this is good to me

KevinMulhern commented on app/javascript/stylesheets/lesson-content.css:
```diff
@@ -24,33 +24,83 @@
+  .lesson-note > *:only-child::before {
+    display: block;
+    font-family: "Font Awesome 6 Free";
+    font-size: 1.8rem;
```
@luuu-xu @thatblindgeye this looks little too big to me in relation to content around it. What do you both think about using something like 1.5? (text-2xl)

1.8
<img width="715" alt="Screenshot 2023-08-09 at 15 06 04" src="https://github.com/TheOdinProject/theodinproject/assets/7963776/d393f7e7-5009-426f-a53f-baa5568edce9">

vs

1.5
<img width="773" alt="Screenshot 2023-08-09 at 15 05 46" src="https://github.com/TheOdinProject/theodinproject/assets/7963776/e2fdd8ac-14b6-4f48-96b3-d200c9d64172">

thatblindgeye replied:
I'm fine with either, as long as the icons are easily recognizable enough. We can always increase/decrease it further if users provide feedback on it

luuu-xu replied:
1.5 and 1.8 both looks fine to me while 1.5 seems more aesthetically pleasant. Let's do 1.5 first.

luuu-xu commented:
@KevinMulhern This time I squashed my commits into one commit and rebased onto just pulled upstream main, and then force pushed to my origin fix branch. Please let me know if it looks good and this approach is okay too? I am lost in the ocean of git commands and practices...

KevinMulhern reviewed (approved):
Nice job with rebasing 😎 it's scary at first, but will become second nature in no time.

Thanks for making those changes, the styles look great and the code looks clean ✨ 
Great work!
